### PR TITLE
Add Quantum Random Walk Tutorial Notebook

### PR DESCRIPTION
**Pull request by Lucaman99**
I created a Jupyter Notebook tutorial on simulating quantum random walks using Cirq, and added it to the examples folder of the repository. The reason there are so many commits to my branch is because the Notebook formatting on Github was a lot different than when I ran the file locally, and thus I needed to do a lot of editing. 😄

Fixes https://github.com/quantumlib/Cirq/issues/1564

### CODE REVIEW

**Comment by vtomole** [comment]
Awesome 😄 . Please add some tests too. For example: https://github.com/quantumlib/OpenFermion-Cirq/blob/master/examples/examples_test.py tests these notebooks: https://github.com/quantumlib/OpenFermion-Cirq/tree/master/examples

**Comment by vtomole** [comment]
I'm looking into https://pypi.org/project/pytest-ipynb/ to see if it would be a better way of testing than the Openfermion-Cirq example.

File: examples/random_walk_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+   "metadata": {},
+   "source": [
+    "# Quantum Random Walks With Cirq - Tutorial"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<i>The first section of this tutorial was taken from this blog post (written by the same person who wrote this tutorial):</i> https://lucaman99.github.io/blog/2019/08/03/Quantum-Random-Walks.html\n",
```
Comment (vtomole) [inline examples/random_walk_tutorial.ipynb]: This sentence is better at the bottom on the tutorial in `Reference` section.

File: examples/random_walk_tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+    "\n",
+    "<i>For more information about quantum random walks, see:</i> https://arxiv.org/abs/quant-ph/0303081"
```
Comment (vtomole) [inline examples/random_walk_tutorial.ipynb]: Put in a reference section at the bottom of the notebook.

File: examples/random_walk_tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Classical Random Walks**\n",
+    "\n",
+    "A random walks is a random process, involving a \"walker\" that is placed in some $n$-dimensional medium (like a grid or a graph). We then repeatedly query some random variable, and based on the outcome of our measurement, the walker's position vector (position on the graph or grid) is updated. A very basic example of a random walk is the one-dimensional graphical case, where we consider a marker placed on the origin of a number line with markings at each of the integers. Let the initial position vector of our marker be $\\biggr\\lvert 0\\rangle$. For $N$ steps of our random walk, so take a set of $N$ random variables, $\\{X_1, \\ ..., \\ X_N\\}$, which can take on either a value of $1$ or $-1$ with equal probability ($0.5$). To find the updated position vector of our walker, we simply compute the value:\n",
```
Comment (vtomole) [inline examples/random_walk_tutorial.ipynb]:
```suggestion
    "A random walk is a random process, involving a \"walker\" that is placed in some $n$-dimensional medium like a grid or graph. We then repeatedly query some random variable, and based on the outcome of our measurement, the walker's position vector on the graph or grid is updated. An example of a random walk is the one-dimensional graphical case, where we consider a marker placed on the origin of a number line with markings at each of the integers. Let the initial position vector of our marker be $\\biggr\\lvert 0\\rangle$. For $N$ steps of our random walk, so take a set of $N$ random variables, $\\{X_1, \\ ..., \\ X_N\\}$, which can take on either a value of $1$ or $-1$ with equal probability ($0.5$). To find the updated position vector of our walker, we simply compute the value:\n",
```

**Comment by vtomole** [comment]
Linking: https://github.com/quantumlib/Cirq/issues/1564

**Comment by Lucaman99** [comment]
Hi @vtomole - I made some of the changes. Sorry, I'm fairly new at using the Github collaboration features, I made the changes that you requested manually, but now I'm realizing that you may have already made the changes and I never accepted the commit. I'm a little bit unclear as to what the Jupyter Notebook tests do and how they should be added into my pull request, I would really appreciate any clarification!

**Comment by vtomole** [comment]
@Lucaman99 You haven't signed the CLA yet.

**Comment by Lucaman99** [comment]
When I go to the link it says I have signed it

**Comment by vtomole** [comment]
@Lucaman99 
 Okay then make a comment "@googlebot I fixed it.." If the bot doesn't comment, it means it doesn't think anything has changed.

**Comment by Lucaman99** [comment]
@googlebot I fixed it

**Comment by Lucaman99** [comment]
@googlebot I fixed it.

**Comment by vtomole** [comment]
@Lucaman99  Is the email in your commits different from the one you used to sign the CLA?

**Comment by Lucaman99** [comment]
@vtomole Yeah, I think so, the CLA won't allow me to change the email

**Comment by vtomole** [comment]
@Lucaman99 I've committed from different emails before with no problems. @Strilanc any tips?

**Comment by Lucaman99** [comment]
@googlebot I fixed it.

File: examples/.ipynb_checkpoints/random_walk_tutorial-checkpoint.ipynb [hunk 4]
```diff
@@ -0,0 +1,766 @@
+{
```
Comment (Strilanc) [inline examples/.ipynb_checkpoints/random_walk_tutorial-checkpoint.ipynb]: This looks like it shouldn't have been included in the commit. Remove it. You can add an `*/.ipynb_checkpoints/*` entry to the `.gitignore` file to ensure it doesn't get recommitted.

**Comment by Lucaman99** [comment]
@vtomole Should I just create another branch and submit a pull request with the correct email?

**Comment by Lucaman99** [comment]
@googlebot I fixed it.

**Comment by vtomole** [comment]
@Lucaman99  Did adding the other email to your Github account not work?

**Comment by Lucaman99** [comment]
No, I don't think so

**Comment by Lucaman99** [comment]
@googlebot I fixed it.

**Comment by vtomole** [comment]
@Lucaman99 No, we can override the cla. You can test by making a "fake" pull request with the "correct" email commit and we'll see if it accepts.

**Comment by Lucaman99** [comment]
@vtomole Yes, I just tried it, and it still didn't seem to work (It said it found a CLA for the author of the pull request, which is my account with the correct email, but can't find one for all the contributors)

**Comment by Strilanc** [comment]
Try squashing everything down into one commit

```
git checkout master -b fresh_branch
git merge old_branch --squash
git commit -m "everything in one commit"
```

You may need to configure your email into git. Check `git config user.email` and `git config --global user.email`.

**Comment by Lucaman99** [comment]
Hey @Strilanc - I read over your review, and I'm so sorry for the massive delay in fixing these issues. I have the bandwidth to fix this hopefully over the next week or so. All of the minor issues will definitely be fixed ASAP. In regards to your overarching criticism, it is totally valid, and I agree. I'll change you structure of the notebook with some of your suggestions. For the multi-controlled-X gates and the arithmetic operations, I'll change them to the new methods, however, I think this is a good opportunity to provide intuition about why these things actually work, and ultimately ensure the person reading this tutorial is confident the quantum walk can be implemented from scratch, with the base set of gates. I'll probably keep the methods I wrote separate from the main quantum walk simulation, and talk a little bit about the decompositions after I introduce the built-in functions. Would that be Ok with you?

**Comment by vtomole** [comment]
@Lucaman99 That's fine to me.

**Comment by Lucaman99** [comment]
I just finished the latest version, will push tonight.

**Comment by Lucaman99** [comment]
Added a PR: #2657 😃

**Comment by vtomole** [comment]
Closing as there is another PR that attempts to accomplish this.

**Pull request closed without merging**